i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,245 @@
+from __future__ import annotations
+
+__all__ = ["ConformerMHSARelPosV1", "ConformerMHSARelPosV1Config"]
+
+from dataclasses import dataclass
+import math
+from typing import Optional, Literal
+
+import torch
+from torch import nn
+import torch.nn.functional as F
+
+from i6_models.config import ModelConfiguration
+from i6_models.util import compat
+from i6_models.parts.dropout import BroadcastDropout
+
+
+@dataclass
+class ConformerMHSARelPosV1Config(ModelConfiguration):
+    """
+    Attributes:
+        input_dim: input dim and total dimension for query/key and value projections, should be divisible by `num_att_heads`
+        num_att_heads: number of attention heads
+        with_bias: whether to add bias to qkv and output linear projections
+        att_weights_dropout: attention weights dropout
+        learnable_pos_emb: whether to use learnable relative positional embeddings instead of fixed sinusoidal ones
+        rel_pos_clip: maximal relative postion for embedding
+        with_linear_pos: whether to linearly transform the positional embeddings
+        separate_pos_emb_per_head: whether to create head-dependent positional embeddings
+        with_pos_bias: whether to add additional position bias terms to the attention scores
+        pos_emb_dropout: dropout for the positional embeddings
+        dropout: multi-headed self attention output dropout
+        dropout_broadcast_axes: string of axes to which dropout is broadcast, e.g. "T" for broadcasting to the time axis
+                                setting to None to disable broadcasting
```

Comment:
I think it would be good to extend the docs, to 1. list what you need to set to get Shawn et al. self-att rel pos 2018 / RETURNN SelfAttentionLayer + RelativePositionalEncodingLayer, and 2. for Dai et al. Transformer-XL style 2019. More or less like the docstring in RETURNN:
```
To get Shawn et al. self-att rel pos 2018 / RETURNN SelfAttentionLayer + RelativePositionalEncodingLayer:

with_bias = False (at least that was the RETURNN behavior)
with_linear_pos = False
with_pos_bias = False
learnable_pos_emb = True
separate_pos_emb_per_head = False (at least that was the RETURNN default)

To get Dai et al. Transformer-XL style 2019:

with_bias = False would be like the paper, however, in most implementations it is True (default)
with_linear_pos = True (default)
with_pos_bias = True (default)
learnable_pos_emb = True False (default)
separate_pos_emb_per_head = True (default)
```

Reply:
I'm not sure about the two general parameters `with_linear_pos` and `separate_pos_emb_per_head`. In my view, their values are not tied to Shaw's or Transformer-XL style relative PE.